EDT.hs

```diff
@@ -455,7 +463,66 @@ module EDT where
+                            ]
+    where predisposition = Distribution [Probability "trustworthy" 0.5
+                                        ,Probability "untrustworthy" 0.5
+                                        ]
+          accuracy       = Distribution [Probability "accurate"   0.99
+                                        ,Probability "inaccurate" 0.01
+                                        ]
+          offer          = Conditional [Clause [Guard "predisposition" "trustworthy",   Guard "accuracy" "accurate"]   "ride"
+                                       ,Clause [Guard "predisposition" "trustworthy",   Guard "accuracy" "inaccurate"] "no ride"
+                                       ,Clause [Guard "predisposition" "untrustworthy", Guard "accuracy" "accurate"]   "no ride"
+                                       ,Clause [Guard "predisposition" "untrustworthy", Guard "accuracy" "inaccurate"] "ride"
+                                       ]
+          location       = Conditional [Clause [Guard "offer" "ride"]    "city"
+                                       ,Clause [Guard "offer" "no ride"] "desert"
+                                       ]
+          action         = Conditional [Clause [Guard "predisposition" "trustworthy", Guard "location" "city"] "pay"
+                                       ,Clause [] "no pay"
+                                       ]
+          value          = Conditional [Clause [Guard "action" "pay",    Guard "location" "city"] "-1000"
+                                       ,Clause [Guard "action" "no pay", Guard "location" "city"] "0"
+                                       ,Clause []                                                 "-1000000"
+                                       ]
+
+  edtChoiceForParfitsHitchhiker = edt [] stdSearch parfitsHitchhiker
+  testEdtChoiceForParfitsHitchhiker = test "EdtChoiceForParfitsHitchhiker" ("pay", -1000) edtChoiceForParfitsHitchhiker
+
+  edtChoiceForParfitsHitchhikerGivenInCity = edt [Guard "location" "city"] stdSearch parfitsHitchhiker
+  testEdtChoiceForParfitsHitchhikerGivenInCity = test "EdtChoiceForParfitsHitchhikerGivenInCity" ("no pay", 0) edtChoiceForParfitsHitchhikerGivenInCity
+
+  cdtChoiceForParfitsHitchhiker = cdt [] stdSearch parfitsHitchhiker
+  testCdtChoiceForParfitsHitchhiker = test "CdtChoiceForParfitsHitchhiker" ("no pay",-990099.0) cdtChoiceForParfitsHitchhiker
+
+  cdtChoiceForParfitsHitchhikerGivenInCity = cdt [Guard "location" "city"] stdSearch parfitsHitchhiker
+  testCdtChoiceForParfitsHitchhikerGivenInCity = test "CdtChoiceForParfitsHitchhikerGivenInCity" ("no pay", 0) cdtChoiceForParfitsHitchhikerGivenInCity
+
+  fdt :: Foldable t => Label -> t Guard -> Search -> Graph Stochastic -> (State, Utility)
+  fdt = stableDT . counterFactualize
+
+  counterFactualize :: Label -> Guard -> Endo [Probability (Graph Deterministic)]
+  counterFactualize l g ps = normalize $ concatMap (\s -> condition g $ intervene (Guard l s) ps) $ choices l ps
```

Comment:
I think the order below makes for an easier explanation:
```haskell
condition g $ nomalize $ concatMap (\s -> intervene (Guard l s) ps) $ choices l ps
```
> We start with a uniform prior over the possible predispositions, then compute the possible universes for each possible intervention on predisposition, and finally condition that distribution on the choice of action we're evaluating.

If we have a monad for a probability distribution we can get rid of normalize and it would look even nicer:
```haskell
condition g $ do predispositionChoice <- uniform $ choices l ps
                 possibleUniverse <- intervene (Guard l predispositionChoice) ps
```

Reply:
@gusbicalho we can try that in the next meetup.